src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/ConvertToJsonCommandV2.cs

```diff
@@ -0,0 +1,610 @@
+                    WriteIndented = !compressOutput,
+
+                    // Set high value to avoid System.Text.Json exceptions
+                    // User-specified depth is enforced by JsonConverterPSObject (max 100 via ValidateRange)
+                    MaxDepth = 1000,
+                    DefaultIgnoreCondition = JsonIgnoreCondition.Never,
+                    Encoder = GetEncoder(stringEscapeHandling),
+                };
+
+                if (enumsAsStrings)
+                {
+                    options.Converters.Add(new JsonStringEnumConverter());
+                }
+
+                // Add custom converters for PowerShell-specific types
+                options.Converters.Add(new JsonConverterInt64Enum());
+                options.Converters.Add(new JsonConverterBigInteger());
+                options.Converters.Add(new JsonConverterNullString());
+                options.Converters.Add(new JsonConverterDBNull());
+                options.Converters.Add(new JsonConverterPSObject(cmdlet, maxDepth));
+                options.Converters.Add(new JsonConverterJObject());
+
+                // Wrap in PSObject to ensure ETS properties are preserved
+                var pso = PSObject.AsPSObject(objectToProcess);
+                return System.Text.Json.JsonSerializer.Serialize(pso, typeof(PSObject), options);
+            }
+            catch (OperationCanceledException)
+            {
+                return null;
+            }
+        }
+
+        private static JavaScriptEncoder GetEncoder(JsonStringEscapeHandling escapeHandling)
+        {
+            return escapeHandling switch
+            {
+                JsonStringEscapeHandling.Default => JavaScriptEncoder.UnsafeRelaxedJsonEscaping,
+                JsonStringEscapeHandling.EscapeNonAscii => JavaScriptEncoder.Default,
+                JsonStringEscapeHandling.EscapeHtml => JavaScriptEncoder.Create(UnicodeRanges.BasicLatin),
+                _ => JavaScriptEncoder.UnsafeRelaxedJsonEscaping,
+            };
+        }
+    }
+
+    /// <summary>
+    /// Custom JsonConverter for PSObject that handles PowerShell-specific serialization.
+    /// </summary>
+    internal sealed class JsonConverterPSObject : System.Text.Json.Serialization.JsonConverter<PSObject>
+    {
+        private readonly PSCmdlet? _cmdlet;
+        private readonly int _maxDepth;
+
+        // Warning tracking (instance variable, reset per serialization)
+        private bool _warningWritten;
+
+        public JsonConverterPSObject(PSCmdlet? cmdlet, int maxDepth)
+        {
+            _cmdlet = cmdlet;
+            _maxDepth = maxDepth;
+        }
+
+        public override PSObject? Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
+        {
+            throw new NotImplementedException();
+        }
+
+        public override void Write(Utf8JsonWriter writer, PSObject pso, JsonSerializerOptions options)
+        {
+            if (LanguagePrimitives.IsNull(pso))
```

Comment:
Really pso can be null.

```suggestion
        public override void Write(Utf8JsonWriter writer, PSObject? pso, JsonSerializerOptions options)
        {
            if (LanguagePrimitives.IsNull(pso))
```

Could you please annotate LanguagePrimitives.IsNull method for nullability?

Reply:
Done in 0e07b33c9. Changed parameter to `PSObject? pso` and annotated `LanguagePrimitives.IsNull` with `[NotNullWhen(false)] object?`.